janus_core/processing/observables.py

```diff
@@ -2,73 +2,254 @@
+            The indices for each self._components.
+        """
+        return [self._allowed_components[c] for c in self.components]
+
+    def _set_components(self, components: list[str]):
+        """
+        Check if components are valid, if so set them.
+
+        Parameters
+        ----------
+        components : str
+            The component symbols to check.
+
+        Raises
+        ------
+        ValueError
+            If any component is invalid.
+        """
+        if any(components - self._allowed_components.keys()):
+            raise ValueError(
+                f"'{components-self._allowed_components.keys()}'"
+                f" invalid, must be '{', '.join(self._allowed_components)}'"
+            )
+
+        self.components = components
 
-class Stress:
+
+# pylint: disable=too-few-public-methods
+class Stress(Observable, ComponentMixin):
     """
     Observable for stress components.
 
     Parameters
     ----------
-    component : str
-        Symbol for tensor components, xx, yy, etc.
+    components : list[str]
+        Symbols for correlated tensor components, xx, yy, etc.
+    atoms_slice : list[int] | SliceLike | None = None
+        List or slice of atoms to observe velocities from.
     include_ideal_gas : bool
         Calculate with the ideal gas contribution.
     """
 
-    def __init__(self, component: str, *, include_ideal_gas: bool = True) -> None:
+    def __init__(
+        self,
+        *,
+        components: list[str],
+        atoms_slice: list[int] | SliceLike | None = None,
+        include_ideal_gas: bool = True,
+    ):
         """
-        Initialise the observables from a symbolic str component.
+        Initialise the observable from a symbolic str component.
 
         Parameters
         ----------
-        component : str
-            Symbol for tensor components, xx, yy, etc.
+        components : list[str]
+            Symbols for tensor components, xx, yy, etc.
+        atoms_slice : list[int] | SliceLike | None = None
+            List or slice of atoms to observe velocities from.
         include_ideal_gas : bool
             Calculate with the ideal gas contribution.
         """
-        components = {
-            "xx": 0,
-            "yy": 1,
-            "zz": 2,
-            "yz": 3,
-            "zy": 3,
-            "xz": 4,
-            "zx": 4,
-            "xy": 5,
-            "yx": 5,
-        }
-        if component not in components:
-            raise ValueError(
-                f"'{component}' invalid, must be '{', '.join(list(components.keys()))}'"
-            )
+        ComponentMixin.__init__(
+            self,
+            components={
+                "xx": 0,
+                "yy": 1,
+                "zz": 2,
+                "yz": 3,
+                "zy": 3,
+                "xz": 4,
+                "zx": 4,
+                "xy": 5,
+                "yx": 5,
+            },
+        )
+        self._set_components(components)
```

Comment:
Does it make sense to make this into a setter, rather than explicitly calling this here (and in `janus_core/processing/observables.py`)?

Reply:
As in instead of the to calls, one to __init__ and one to _set_components, just having one setter? If so we might as well roll it into one

Reply:
right I see what you mean

Reply:
@ElliottKasoar I've made ```self.components``` a property with a ```@components.setter``` so we have this behaviour/style now